### DISCUSSION TITLE
command line-specific problem with loading neutral losses from MSFragger Lib?

### DISCUSSION THREAD

**Discussion by JM-Bader** [vdemichev/DiaNN#377] [Q&A] [answered]
I run DIA-NN with a library on the command line tool and get the following warning (in the log): 
No neutral loss information found in the library. 

![image](https://user-images.githubusercontent.com/98888664/167827403-c78fbe80-d8e8-4f7b-aa16-603c4fcb2b8d.png)

The library was generated with MSFragger. I am no MSFragger expert and I did not find any button in fragpipe to enable/disable neutral losses in MSFragger, so I assume MSFragger does routinely include neutral losses in its searches and also includes them in the library. 

Interestingly, when using the very same library in a DIA-NN GUI search, the warning does not appear: 

![image](https://user-images.githubusercontent.com/98888664/167827920-5db176a4-db45-472e-8401-a76ea49c337a.png)

So: 
Q1: Is there a discrepancy between GUI and command-line?
Q2: Could it be that there is a loading issue specifically in the command-line version?
Q3: Could it be that the GUI just doesn't give the warning?
Q4: What to do to fix it?
Q5: Anyway, what is your recommended library-generation tool ? (Spectronaut, MSFragger/fragpipe, MQ, other?)

**Comment by vdemichev** [comment]
Q1: this is strange. Can you please copy-paste here the column names of the library? 
Q2: it might be that the issue is under Linux. 
Q3: No.
Q4: This will not affect anything, unless you use MBR with 'Smart profiling' or (not recommended) 'Full profiling'.
Q5: MSFragger or, if from DIA data, DIA-NN itself. Also can use lib-free mode.

**Reply by JM-Bader** [reply]
Q1: Here is the column names of the library: 
![image](https://user-images.githubusercontent.com/98888664/168090267-0c932afa-dbde-46ac-8f0c-6e281db39bc7.png)
Apparantly, there are only NANs in the fragment loss type....
However the GUI does report a warning. 
Q2: shouldn't the GUI likewise report a warning. Isn't that concerning?
Q3: Do I have to use special MSFragger settings to get a compatible library with neutral loss information?

**Comment by vdemichev** [comment, marked as answer]
There should be no warning. NAs are interpreted as unknown losses, which are ignored. 
But on Linux for whatever reason it does not see the last column in the library, which causes this. Workaround is probably to add another column to the library, maybe it will solve it. Specifically with this library, as there are no losses in it, it's not a problem though.

**Reply by vdemichev** [reply]
I will look into this.

**Reply by JM-Bader** [reply]
ok, great thanks. 
Q4: Do I understand correctly that no loss is meant to be encoded as NAN (as opposed to sth else more explicit like 'noloss'). That way is not clear whether the data is actually missing in the table due to an issue or that there just 0 losses. 
Q5: Isn't it very unlikely that 0 fragments in the library have a neutral loss? (Independently, there could be a MSFragger issue that leads to all these NANs...)

**Reply by vdemichev** [reply]
It should be 'noloss', but NA will be interpreted the same way.
If MSFragger does not consider neutral losses, then there will be none in the lib

**Reply by fcyu** [reply]
Correct. MSFragger (FragPipe more precisely, because MSFragger is just a search engine, you need other tools in FragPipe to filter the PSMs and build the spectral library) does not consider neutral loss. We will add neutral loss info in the future release.

@vdemichev , could you please tell me what info should be in that column? A small example would be good enough.

Thanks,

Fengchao

**Reply by vdemichev** [reply]
'noloss'

**Reply by fcyu** [reply]
What if the fragment has a neutral loss?

**Reply by JM-Bader** [reply]
thanks for your support. 
I had just raised this in the MSFragger github: https://github.com/Nesvilab/MSFragger/issues/223
@fcyu It appears very unlikely to me that this is just about labelling it as noloss. I would expect some fragments to have a loss.

**Reply by vdemichev** [reply]
This is the decoding code 
```
if (wl == "noloss") p.loss = loss_none;
									else if (wl == "H2O") p.loss = loss_H2O;
									else if (wl == "NH3") p.loss = loss_NH3;
									else if (wl == "CO") p.loss = loss_CO;
									else if (wl == "H3PO4") p.loss = loss_H3PO4;
									else if (wl == "H4COS") p.loss = loss_H4COS;
									else p.loss = loss_other;
```

**Reply by fcyu** [reply]
Thanks @vdemichev , we will adopt the code.

Best,

Fengchao

**Reply by JM-Bader** [reply]
> There should be no warning. NAs are interpreted as unknown losses, which are ignored. But on Linux for whatever reason it does not see the last column in the library, which causes this. Workaround is probably to add another column to the library, maybe it will solve it. Specifically with this library, as there are no losses in it, it's not a problem though.

I added another column as a workaround. 
![image](https://user-images.githubusercontent.com/98888664/168113732-3bda14f6-1622-45fe-b907-705388a3852e.png)

When running the DIA-NN command line search now, there are about 30% less proteins in the library
Normal:
![image](https://user-images.githubusercontent.com/98888664/168114289-6d8515b0-ffcb-48ac-bd47-bb56df8941b6.png)


With extra column:
![image](https://user-images.githubusercontent.com/98888664/168114162-3304a93f-bdd8-430d-b40c-c0beba1b250b.png)

Q1: Could the extra column mess up the reading-in of the library file etc?
Q2: If its just about the warning, then its ok to just get the warning (as of now) and use the old version of the library - since MSFragger doesn't incldue neutral loss fragments. That still holds true i guess, correct?
Q3: What about the other comments you made about using MBR in conjunction with smart profiling (recommended) or full profiling (not recommmended) which would lead to problems? I do want to use MBR. What should I use then for library generation settings in DIA-NN? And why does it matter ? What is the link to the neutral loss isse with MSFragger?

**Reply by fcyu** [reply]
> thanks for your support. I had just raised this in the MSFragger github: https://github.com/Nesvilab/MSFragger/issues/223 @fcyu It appears very unlikely to me that this is just about labelling it as noloss. I would expect some fragments to have a loss.

If you click the link you posted, you will be redirected to the issue (which has been transferred to FragPipe repo).

**Reply by vdemichev** [reply]
Something is clearly wrong with library formatting. Maybe some extra symbol somewhere. Can you please attach this library you use under Linux?

Vadim

**Reply by JM-Bader** [reply]
> Something is clearly wrong with library formatting. Maybe some extra symbol somewhere. Can you please attach this library you use under Linux?
> 
> Vadim

these libraries are confidential. I will send you a private email with them
